sources/models/src/modeled_types/shared.rs

```diff
@@ -349,3 +349,77 @@ mod test_version {
         }
     }
 }
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// DNSDomain represents a string that is a valid DNS domain. It stores the
+/// original string and makes it accessible through standard traits. Its purpose
+/// is input validation, for example validating the kubelet's "clusterDomain"
+/// config value.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct DNSDomain {
+    inner: String,
+}
+
+impl TryFrom<&str> for DNSDomain {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
```

Comment:
I'm wondering if it would be simpler / more resilient to use `url::Host::parse` on the input string and verify the result is the `Host::Domain` variant (or something similar with another library, but we already have `url`) rather than parsing and implementing domain name checking rules ourselves.  Then, if there are further restrictions that Kubernetes has, we could check those afterward.  Or is this already pretty specific to what Kubernetes will accept?

(My comments on the implementation wouldn't apply if we change this approach)

Reply:
The kubelet doesn't do any validation on the clusterDomain setting. It uses whatever string is supplied. I wanted to add some basic sanity checking.

Using url::Host:parse to validate the clusterDomain makes sense to me. I'll make the change. I'm new to rust and wasn't sure what the best practice was for validating a DNS domain.